rcl_action/src/rcl_action/action_server.c

```diff
@@ -641,12 +641,16 @@ rcl_action_expire_goals(
     }
     goal_time = _goal_info_stamp_to_nanosec(info_ptr);
     if ((current_time - goal_time) > timeout) {
-      // Stop tracking goal handle
-      // Fill in any gaps left in the array with pointers from the end
+      // Deallocate space used to store pointer to goal handle
+      allocator.deallocate(action_server->impl->goal_handles[i], allocator.state);
+      action_server->impl->goal_handles[i] = NULL;
+      // Move all pointers after backwards one to fill the gap
+      for (size_t post_i = i; (post_i + 1) < num_goal_handles; ++post_i) {
+        action_server->impl->goal_handles[post_i] = action_server->impl->goal_handles[post_i + 1];
+      }
+      // decrement i to check the same index again now that it has a new goal handle
+      --i;
       --num_goal_handles;
-      action_server->impl->goal_handles[i] = action_server->impl->goal_handles[num_goal_handles];
-      allocator.deallocate(action_server->impl->goal_handles[num_goal_handles], allocator.state);
```

Comment:
Would the logic also be fixed if this deallocate was for index `i` and moved one line up?
For sure the `--i;` is needed.
```
allocator.deallocate(action_server->impl->goal_handles[i], allocator.state);
action_server->impl->goal_handles[i] = action_server->impl->goal_handles[num_goal_handles];
action_server->impl->goal_handles[num_goal_handles] = NULL;
// decrement i to check the same index again now that it has a new goal handle
--i;
```
Just want to make sure I understand the bug :)

Reply:
That would be an improvement, though it would still have a bug. All goal handles need to be moved to shrink the array, but this shrinking logic is doing it one at a time in a block that only executes if the goal handle at `i` has expired. An active goal at the end of the array would be lost when the array shrinks because this bit wouldn't be reached. If there were two goals that expired at once then goal handles after the second expired one would only be moved back one place when they need to be moved back two places.

This PR adds a for-loop to make sure all goals after the expired one are moved.

Reply:
It makes sense to me, thanks for the explanation!